lib/manageiq/api/common/user.rb

```diff
@@ -0,0 +1,63 @@
+module ManageIQ
+  module API
+    module Common
+      class User
+        IDENTITY_KEY = 'x-rh-identity'.freeze
+        VALID_USER_HEADER_KEYS = %w(account_number user username email is_org_admin)
+
+        def self.with_logged_in_user(user = nil)
+          yield Thread.current[:attr_current_user] = new(user)
+        ensure
+          Thread.current[:attr_current_user] = nil
+        end
+
+        def initialize(user = nil)
+          @current = user || ManageIQ::API::Common::Headers.current
+          user_hash = validate
+          @identity = user_hash['identity']
+          build_methods
+          self
+        end
+
+        private
+
+        def decode(key = IDENTITY_KEY)
+          raise StandardError, "Requires a valid header or user hash" unless @current
+          if @current.is_a?(ActionDispatch::Http::Headers)
+            val = @current.instance_variable_get(:@req).stringify_keys
+          else
+            val = @current.stringify_keys
+          end
+          JSON.parse(Base64.decode64(val[key]))
+        end
+
+        def validate
+          check_user(decode)
+        end
+
+        def check_user(user)
```

Comment:
The validation code is a little hard to read because the VALID_USER_HEADER_KEYS is a mix of top level and second level keys. 
If we explicitly construct attribute methods we can move validation to each individual method. Raise an error if a mandatory attribute is missing; return `nil` if an optional attribute is missing.